https://issuehunt.io/repos/114969039/issues/8

june07 opened the pull request:
Added Google Cloud Storage support

- Upload backup data to Google Cloud Storage
- Restore from Google Cloud Storage
- Manage data on Google Cloud Storage
- Delete expired files
- Add environment variables to configure
- Modify README.md

All items have been completed.

Commit:
Added support for Google Cloud Storage.  https://github.com/weseek/mongodb-awesome-backup/issues/8

june07 committed:
Saved README.md with Dillinger.io

june07 committed:
Saved README.md with Dillinger.io

june07 committed:
Saved README.md with Dillinger.io

june07 committed:
Saved README.md with Dillinger.io

ryu-sato commented on bin/backup.sh:
```diff
@@ -32,11 +33,16 @@ TARBALL_FULLPATH="${TMPDIR}/${TARBALL}"
-  echo "ERROR: The environment variable S3_TARGET_BUCKET_URL must be specified." 1>&2
+# deprecate the old option
+if [ "x${S3_TARGET_BUCKET_URL}" != "x" ]; then
+  TARGET_BUCKET_URL=$S3_TARGET_BUCKET_URL
```
👍 
Could you add print message like "WARNING: The environment variable S3_TARGET_BUCKET_URL is deprecated." to announce to users.

june07 replied:
Done.

ryu-sato commented on README.md:
```diff
@@ -62,7 +73,10 @@ You can use "**restore**" command to restore database from backup file.
+  -e GCP_SECRET_ACCESS_KEY=<Your GCP Secret> \
+  -e GCP_PROJECT_ID=<Your GCP Project ID> \
+  -e TARGET_BUCKET_URL=<Target Bucket URL ([s3://...|gs://...])> \
   -e S3_TARGET_FILE=<Target S3 file name to restore> \
```
`S3_TARGET_FILE` seem to be changed to `TARGET_FILE`, right?
Please modify this.

june07 replied:
Done.

ryu-sato commented on test/docker-compose.yml:
```diff
@@ -27,7 +27,7 @@ services:
       <<: *base_env
-      S3_TARGET_BUCKET_URL: s3://app_restore/
+      TARGET_BUCKET_URL: s3://app_restore/
       S3_TARGET_FILE: dummy-backup-20180622.tar.bz2
```
Please modify `S3_TARGET_FILE` to `TARGET_FILE`.

june07 replied:
Done.

ryu-sato reviewed (changes requested):
1. Please modify `README.md`, `bin/backup.sh`, and ` test/docker-compose.yml`. I was commented the point of modification.
2. The PR is implemented to do with HMAC authentication, but GCS prefered authentication with OAuth 2. ([ref](https://cloud.google.com/storage/docs/gsutil/addlhelp/CredentialTypesSupportingVariousUseCases))
Most of the tasks of the issue are completed, but could you implement with OAuth2?
Maybe the option of `docker run` will differ from S3 pattern like below.


```
$ docker run -it --net growidockercompose_default --rm -e TARGET_BUCKET_URL=gs://***/ weseek/mongodb-awesome-backup
This command will create a boto config file at /root/.boto containing
your credentials, based on your responses to the following questions.
Please navigate your browser to the following URL:
https://accounts.google.com/o/oauth2/auth?scope=***
In your browser you should see a page that requests you to authorize access to Google Cloud Platform APIs and Services on your behalf. After you approve, an authorization code will be displayed.

Enter the authorization code: **************

Please navigate your browser to https://cloud.google.com/console#/project,
then find the project you will use, and copy the Project ID string from the
second column. Older projects do not have Project ID strings. For such projects,
 click the project and then copy the Project Number listed under that project.

What is your project-id? ryutest-99285

gsutil developers rely on user feedback to make improvements to the
tool. Would you like to send anonymous usage statistics to help
improve gsutil? [y/N]

Boto config file "/root/.boto" created. If you need to use a proxy to
access the Internet please see the instructions in that file.
=== ./backup.sh started at 2019/02/19 19:26:56 ===
dump MongoDB...
2019-02-19T19:26:56.590+0900    writing admin.system.version to
  : <snip>
```

ryu-sato commented:
> 2. The PR is implemented to do with HMAC authentication, but GCS prefered authentication with OAuth 2. ([ref](https://cloud.google.com/storage/docs/gsutil/addlhelp/CredentialTypesSupportingVariousUseCases))
>     Most of the tasks of the issue are completed, but could you implement with OAuth2?
>     Maybe the option of `docker run` will differ from S3 pattern like below.
> ```
> $ docker run -it --net growidockercompose_default --rm -e TARGET_BUCKET_URL=gs://***/ weseek/mongodb-awesome-backup
> ```

Above example was not good.

We thought better example is that user create `.boto` by `gsutil` at first, and next execute docker container with mounting `.boto` file like below.

```
$ cd /path/to/mongodb-awesome-backup-repository/
$ gsutil config -o gsutil-config/.boto # Create .boto file at first. (*1)
$ docker run -it  --net growidockercompose_default -v $(pwd)/gsutil-conf/.boto:/root/.boto --rm -v config/.boto:/root/.boto -e TARGET_BUCKET_URL=gs://***/ weseek/mongodb-awesome-backup
=== ./backup.sh started at 2019/02/22 18:59:06 ===
dump MongoDB...
  : <snip>
```
(*1) It is needed to install gsutil in host of docker container.

Finally we expected to **REMAIN** implementation with HMAC authentication.

Commit:
Added deprecation warning message.  Changed S3_TARGET_FILE to TARGET_FILE.  Changed `S3_TARGET_FILE` to `TARGET_FILE`

june07 commented:
The reason I chose to go with HMAC authentication was to maintain parity with the current AWS S3 implementation.  Although oauth is the recommended method, it requires input from the user at runtime which is different from the current behavior of the S3 code.  If oauth is used then the user will be prompted to provide input...

Commit:
Merge branch 'master' of https://github.com/weseek/mongodb-awesome-backup

ryu-sato commented:
> 1. Please modify `README.md`, `bin/backup.sh`, and ` test/docker-compose.yml`. I was commented the point of modification.

👍 OK.

> The reason I chose to go with HMAC authentication was to maintain parity with the current AWS S3 implementation. Although oauth is the recommended method, it requires input from the user at runtime which is different from the current behavior of the S3 code. If oauth is used then the user will be prompted to provide input...

Yes...

I think implementing GCS with HMAC authentication is a good simple idea.
But enable to OAuth is a better idea.

I think all we needs to enable OAuth is that execute `boto.sh` when `/root/.boto` is not exists.
So, I think that it will not force a heavy load to user and to you.

Commit:
Added auth0 support.

june07 committed:
Saved README.md with Dillinger.io

Commit:
Fixed testing fail when .bota does not exist.

june07 commented:
Added auth0 support and updated the README to reflect the changes.  Works nicely in that user input is only required on the first run and the .boto is saved to the user home directory (can be changed easily).

Commit:
Merge branch 'master' of github.com:june07/mongodb-awesome-backup

Commit:
Exit bota.sh code alltogether if we are using AWS.

Commit:
Turned off verbose logging.

skomma commented on README.md:
````diff
@@ -12,20 +12,31 @@ Requirements
+
+* Google Cloud Interoperable storage access keys (see https://cloud.google.com/storage/docs/migrating#keys)
+  * GCP_ACCESS_KEY_ID, GCP_SECRET_ACCESS_KEY, and GCP_PROJECT_ID are only required if using HMAC authentication.
+  * When using oauth authentication, a docker mount ``` -v ~:/mab``` can be added to save auth0 credentials to your home directory after mongodb-awesome-backup is run.  On subsequent runs, the same ```~/.bota``` file will be used for authentication.
````
`` ``` `` may be `` ` ``?

And, what are `mab` and `bota`?
Are there strings samples?

june07 replied:
Replaced ``` with `.  "mab" is the name of the mount point inside of the Docker container, pretty standard stuff.

skomma commented on README.md:
```diff
@@ -12,20 +12,31 @@ Requirements
-  -e S3_TARGET_BUCKET_URL=<Target S3 Bucket URL (s3://...)> \
+  [ -e GCP_ACCESS_KEY_ID=<Your GCP Access Key> \
+  -e GCP_SECRET_ACCESS_KEY=<Your GCP Secret> \
+  -e GCP_PROJECT_ID=<Your GCP Project ID> \ ]
```
Could you add `[]` in each line?

skomma commented on README.md:
```diff
@@ -42,15 +53,19 @@ and after running this, `backup-YYYYMMdd.tar.bz2` will be placed on Target S3 Bu
+  [ -e GCP_ACCESS_KEY_ID=<Your GCP Access Key> \
+  -e GCP_SECRET_ACCESS_KEY=<Your GCP Secret> \
+  -e GCP_PROJECT_ID=<Your GCP Project ID> \ ]
+  -e TARGET_BUCKET_URL=<Target Bucket URL ([s3://...|gs://...])> \
```
Could you add `[]` in each line?

skomma commented on README.md:
```diff
@@ -62,14 +77,18 @@ You can use "**restore**" command to restore database from backup file.
+  -e GCP_SECRET_ACCESS_KEY=<Your GCP Secret> \
+  -e GCP_PROJECT_ID=<Your GCP Project ID> \ ]
+  -e TARGET_BUCKET_URL=<Target Bucket URL ([s3://...|gs://...])> \
+  -e TARGET_FILE=<Target S3 or GS file name to restore> \
```
Could you add `[]` in each line?

june07 replied:
Added.

skomma reviewed (changes requested):
I have a few questions, and could you fix a few points?

june07 committed:
Saved README.md with Dillinger.io

june07 committed:
Saved README.md with Dillinger.io

skomma commented on README.md:
```diff
@@ -12,20 +12,31 @@ Requirements
+
+* Google Cloud Interoperable storage access keys (see https://cloud.google.com/storage/docs/migrating#keys)
+  * GCP_ACCESS_KEY_ID, GCP_SECRET_ACCESS_KEY, and GCP_PROJECT_ID are only required if using HMAC authentication.
+  * When using oauth authentication, a docker mount ` -v ~:/mab` can be added to save auth0 credentials to your home directory after mongodb-awesome-backup is run.  On subsequent runs, the same `~/.bota` file will be used for authentication.
```
Perhaps, is `mab` the abbreviation of "Mongodb Awesome Backup"?

And is `~/.bota` mistake of `~/.boto`??

june07 replied:
Ha, ha!  Yes, you are right... bota should be boto!  Fixed it.  I did that more than once during the execution of this PR.  As you're 2 for 2, as mab is the abbreviation of "Mongodb Awesome Backup".  Cheers!

skomma replied:
Hmm...I think users cannot read the meaning of "mab" from this document like me...

Could you add an explanation of "mab"?
This is the final change request.

june07 replied:
Done

ryu-sato reviewed (approved):
> Added auth0 support and updated the README to reflect the changes. Works nicely in that user input is only required on the first run and the .boto is saved to the user home directory (can be changed easily).

Good Job! approve.

june07 committed:
Saved README.md with Dillinger.io

june07 committed:
Saved README.md with Dillinger.io
